### ISSUE TITLE
"Text file busy" error when kata-deploy runs on existing kata-deploy deployment

### ISSUE THREAD

**Issue by Jc2k** [kata-containers/packaging#898]
# Description of problem

"Text file busy" errors when running kata-deploy on a system where kata containers are already running

# Expected result

Files are replaced - maybe `cp -f` or `cp --remove-destination`?

# Actual result

From logs:

```
copying kata artifacts onto host
cp: cannot create regular file '/opt/kata/bin/qemu-system-x86_64': Text file busy
cp: cannot create regular file '/opt/kata/libexec/kata-containers/kata-shim': Text file busy
cp: cannot create regular file '/opt/kata/libexec/kata-containers/kata-proxy': Text file busy
```

**Comment by grahamwhaley** [comment]
/cc @egernst @amshinde 
I guess there are two potential questions here:
1) Is that an expected fail case, and can/should we fail more gracefully (`Abort: already installed` for instance).
2) Could this be, or is there already, a node upgrade method for `kata-deploy`?

**Comment by Jc2k** [comment]
Whilst i raised the ticket as though i was upgrading i guess i should be clear that i'm not. The container was restarted for whatever reason and now its looping (802 attempts so far). So idempotence or more error handling is the main thing for me.

As an end user, my ideal stretch goal would be that i could point the daemonset image at a new tag to initiate a rolling release of a new kata version over a bunch of nodes. That would be ace. I'm still evaluating kata-deploy (over the apt repo) so don't know if this is the intention or not.

**Comment by egernst** [comment]
Hey @Jc2k  -- I agree with your assessment re: relative priority. Rolling update would be good, but seems the node would have to not be running any instances as is.  Not ideal.

**Comment by Jc2k** [comment]
Yes, avoiding the race that leaves me with kata-deploy pods in restart loops would definitely be the priority for me.

As for rolling update - pretty sure we can do it without stopping pods - i suspect `cp --remove-destination` would work as you are allowed to delete a file that is a running executable, just not write to it in place. You could probably do a rename to make it atomic - potentially copy all the binaries into an adjacent folder and rename the old folder over the new folder, which has the added benefit of avoiding a small window where a pod starting might see some files from one version of kata and some from another.

**Comment by jodh-intel** [comment]
@egernst - could you tal at this. I think calling `install` rather than `cp` would resolve the immediate issue?

**Comment by Jc2k** [comment]
@jodh-intel from my point of view this issue was never resolved and the problem still exists in the new location for kata-deploy (https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/scripts/kata-deploy.sh#L52).

At the moment I have a fork of kata-deploy that uses `rsync` rather than `cp` to make kata-deploy not fail.

(I've also had to remove the preStop lifecycle as that had a habbit of restoring containerd's config from a stale backup reverting changes made by something else, but I was waiting to resolve this ticket before bringing that up).

**Comment by fidencio** [comment]
@Jc2k, please, see http://lists.katacontainers.io/pipermail/kata-dev/2021-April/001819.html, there you'll find the explanation why the issue was closed.

**Comment by fidencio** [comment]
```
Hi All,

# TL;DR

We're closing all open GitHub issues for Kata 1.x and 2.x that were created before 1st June 2020.


# Background 

We have too many old open issues that are making it hard to identify the reallly important newer ones.

# Plan

Since a lot of those issues are now probably "stale", we intend to close all open GitHub issues opened before 1 June 2020.

# When?

This week some time.

# Why?

Closing these old issues will:

- Allow us to better manage the more recent issues

  We're in the process of categorising current issues with labels to allow us to manage the backlog more efficiently).

- Act as a forcing function for those who raised the issue

  The originator will get notified when the issue is closed, which will remind them of the issue and the problem. And if the issue is still important, they can re-open it with a single click.

# What is the impact of closing an issue?

Whether a GitHub issue is open or closed is simply a state field - you can open and close an issue as many times as you like: no information is going to be lost by changing the issue state to closed. By closing the old issues, we'll hopefully get the list of relevant open issues down to a more manageable number. And once labelled, will allow us to concentrate on fixing the issues rather than spending lots of time managing the issues themselves.

Of course, if an issue is closed, it's highly unlikely anyone will be working on resolving it, so...

# If one of your issues is closed

As mentioned above, if one of your issues is closed, please consider whether you are happy with this. If you think the issue should remain open, just click the "reopen" button and add a comment for us. It would be helpful if you could also mention whether the problem still exists with the latest 2.x version of Kata.

# Get involved

If you're interested in helping us manage the isue backlog, please get in contact [1], and maybe consider volunteering to help on the Kata rota [2].

Cheers,

James

[1] - https://github.com/kata-containers/community/blob/master/README.md#join-us
[2] - https://github.com/kata-containers/community/wiki/Review-Team-Rota
```

**Comment by Jc2k** [comment]
@fidencio thanks for taking the time to post that! Unfortunately I don't seem to have an option to "Reopen" this ticket, though I would very much like to. I can confirm it still exists in 2.x.

**Comment by fidencio** [comment]
@Jc2k, I can re-open it t you, but I'd like to also bring this to your attention: http://lists.katacontainers.io/pipermail/kata-dev/2020-November/001601.html

```
Hey kata folk --

With 2.0 out now, we are focusing continued development and improvements on
the 2.0 code base.  As a result, in a prior architecture committee meeting
discussion, we made the decision on a 6 month window (from release of 1.12)
before stopping regular support of the 1.x (1.11 and 1.12 currently)
branches.

Based on this, here's a first heads up!  We plan to deprecate support for
1.x on May 12th 2021.

Thanks!
Eric
```

So, I'd very much appreciate if you could open an issue on https://github.com/kata-containers/kata-containers/issues/new/choose and we can treat that there.

If you also want to have this one re-opened, let me know and I'll do it.

**Comment by fidencio** [comment]
Last but not least, just taking a look here, it seems to be a simple patch.  So, I'm re-opening this now and will be waiting for the 2.x issue to be filed.

@Jc2k, thanks for contacting us!

**Comment by fidencio** [comment]
@Jc2k, would you be interested on submitting a patch?

In any case, for whoever works on this in the future, basically, what we need to do is simply avoid using `cp -a` and use `install` instead, in the following parts:
https://github.com/kata-containers/packaging/blob/17f37e029c2fe20f7d514e6e3a01cfb9a2523aaa/kata-deploy/scripts/kata-deploy-docker.sh#L29
https://github.com/kata-containers/packaging/blob/17f37e029c2fe20f7d514e6e3a01cfb9a2523aaa/kata-deploy/scripts/kata-deploy.sh#L53

**Comment by Jc2k** [comment]
@fidencio Ah, i assumed you would use the "Transfer issue" function if needed so we didn't loose the existing discussion between the team. Let me know if you can't use the "Transfer issue" button or if its not appropriate for project workflow reasons and i'll happily open another one and close this (or even leave both open, whatever works for you guys).

I personally am only interested in the 2.x version of the bug at this point, but as the shell script is mostly the same I guess it can be fixed in both places.

I tried to use `install` as was suggested but it didn't work out (i believe because there were some directories and `install` isn't recursive). `rsync` has been working well in my environment since i raised this, but that did mean actually installing rsync from yum - i wasn't sure how palatable that was for this project.

As you say, if the extra package is ok its a 2 line patch.
